cmd/charm-ingest/whitelist.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/charmstore-client/internal/ingest"
+	"gopkg.in/juju/charmrepo.v4/csclient/params"
+)
+
+var validChannels = params.ValidChannels
+
+// parseWhitelist takes an io.Reader for a whitelist and returns []ingest.WhitelistEntity
+// or an error if unable to parse
+func parseWhitelist(r io.Reader) ([]ingest.WhitelistEntity, error) {
+	var entities []ingest.WhitelistEntity
+	fileScanner := bufio.NewScanner(r)
+	lineNum := 0
+
+	for fileScanner.Scan() {
+		lineNum++
+
+		entity, err := parseLine(fileScanner.Text(), lineNum)
+		if err != nil {
+			return nil, err
```

Comment:
We tend to wrap errors rather than returning them bare.

    return nil, errgo.Mask(err)

Or how about this?

    return nil, errgo.Notef(err, "line %d", lineNum)

Then you won't need to pass lineNum into parseLine.